### PR TITLE
Introduce `default_args` context manager

### PR DESCRIPTION
**Pull request by haampie**
This adds a rather trivial context manager that lets you deduplicate repeated
arguments in directives, e.g.

```python
depends_on("py-x@1", when="@1", type=("build", "run"))
depends_on("py-x@2", when="@2", type=("build", "run"))
depends_on("py-x@3", when="@3", type=("build", "run"))
depends_on("py-x@4", when="@4", type=("build", "run"))
```

can be condensed to

```python
with default_args(type=("build", "run")):
    depends_on("py-x@1", when="@1")
    depends_on("py-x@2", when="@2")
    depends_on("py-x@3", when="@3")
    depends_on("py-x@4", when="@4")
```

The advantage is it's clear for humans, the downside it's less clear for type checkers due to type erasure

### CODE REVIEW

**Review by tgamblin — CHANGES_REQUESTED** [review]
This looks awesome to me, but can you make it so that `with when` is implemented using the same mechanism?  i.e., shouldn’t this:

```python
with when(“@2:”):
```

become syntactic sugar for:

```python
with default_args(when=“@2:”):
```
?

I think `when` is special enough to keep, and it’s more readable for its special case. But we don’t need two code paths in the implementation for basically the same thing.

**Comment by haampie** [comment]
Hm, it's not exactly the same, since `when` combines constraints, while `default_args` overrides.

What's expected here:

```python
with default_args(type="build"), default_args(type=("link", "run")):
    depends_on("x")
```

is that equivalent to

```python
depends_on("x", type=("build", "link", "run"))
````

or

```python
depends_on("x", type=("link", "run"))
```

**Comment by alalazo** [comment]
@tgamblin `when` deals with a single specific argument, that is known to be a spec literal:
1. It combines constraints using `Spec.constrain`
2.  It can be nested (though we don't recommend that in general), and 
3. The `when=` argument can be in the directive too. 

This:
```python
with when("+foo"):
    with when("+bar"):
        depends_on("xxx", when="+baz")
```
is equivalent to:
```python
depends_on("xxx", when="+foo+bar+baz")
```

**Comment by tgamblin** [comment]
Ok, I agree with @alalazo and that's a great reason why `when` should have its own directive.

Are there any other default args worth having special semantics for?  I can't think of any off the top of my head that I'd justify implementing semantics other than override for.

but RE: @haampie, this is awful:

```python
with default_args(type="build"), default_args(type=("link", "run")):
    depends_on("x")
```

I think most people would expect that to be equivalent to `depends_on("x", type=("build", "link", "run"))` as written, but then you should just write that. Sigh.  I guess we could try to catch that in audit or review.

I could see you wanting an override if the second context manager were in the middle of some statements like:

```python
with default_args(when="@2.0:", type="build"):
    depends_on("x")
    with default_args(type=("link", "run")):
        depends_on("y")
        depends_on("z")
```

but I think we should just encourage people not to do this.  I worry about people composing `with when()` and `with default_args()` as it's going to be surprising if they use the latter for `when`.  e.g. what should these do and what do they do right now?

```python
with default_args(when="@2.0:", type="build"):
    depends_on("x")
    with when("@3.0:"):
        depends_on("y")
        depends_on("z")
```

or:

```python
with default_args(when="@2.0:", type="build"):
    depends_on("x")
    with when("@3.0:"):
        depends_on("y")
        depends_on("z")
```

Should this be disallowed?  I kind of think we should just raise and recommend using `with when` if you try to stick `when` in `default_args`.

Finally, I think regardless of what we settle on, the semantics need to be documented with a note saying how they're different from `with when`.

**Comment by adamjstewart** [comment]
> Are there any other default args worth having special semantics for?

Would love to be able to use `with type(...)` but I imagine that's not desirable for the same reasons that `with args(...)` isn't 😆

**Comment by haampie** [comment]
@tgamblin should be good to go? For documentation see https://spack--39964.org.readthedocs.build/en/39964/packaging_guide.html#common-default-arguments

**Comment by haampie** [comment]
And this one too @tgamblin :D

**Review by tgamblin — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-black/package.py [hunk 1]
```diff
@@ -37,23 +37,24 @@ class PyBlack(PythonPackage):
...
+        depends_on("py-platformdirs@2:")
+        depends_on("py-tomli@1.1:", when="@22.8: ^python@:3.10")
+        depends_on("py-tomli@1.1:", when="@21.7:22.6")
+        depends_on("py-typing-extensions@3.10:", when="^python@:3.9")
+
+    depends_on("py-colorama@0.4.3:", when="+colorama")
+    depends_on("py-uvloop@0.15.2:", when="+uvloop")
+    depends_on("py-aiohttp@3.7.4:", when="+d")
+    depends_on("py-ipython@7.8:", when="+jupyter")
+    depends_on("py-tokenize-rt@3.2:", when="+jupyter")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-black/package.py:57]: These should be indented...

**Pull request merged**